File: Arrays/03_ContainsDuplicate/solution.cpp
/*
Problem 3: Contains Duplicate
Given an integer array nums, return true if any value appears at least twice.
Time Complexity: O(n), Space Complexity: O(n)
  */

#include <iostream>
#include <vector>
#include <unordered_set>
using namespace std;

class Solution {
public:
    bool containsDuplicate(vector<int>& nums) {
              unordered_set<int> seen;
        for (int num : nums) {
            if (seen.count(num)) return true;
            seen.insert(num);
        }
        return false;
    }
};

int main() {
      Solution sol;
    vector<int> nums1 = {1,2,3,1};
    vector<int> nums2 = {1,2,3,4};
    cout << boolalpha << sol.containsDuplicate(nums1) << endl;  // true
    cout << sol.containsDuplicate(nums2) << endl;  // false
    return 0;
}

File: Arrays/03_ContainsDuplicate/solution.py
"""
Problem 3: Contains Duplicate
Given an integer array nums, return true if any value appears at least twice,
and return false if every element is distinct.

Time Complexity: O(n)
Space Complexity: O(n)
"""

from typing import List

def containsDuplicate(nums: List[int]) -> bool:
      seen = set()
      for num in nums:
                if num in seen:
                              return True
                          seen.add(num)
            return False


if __name__ == "__main__":
      print(containsDuplicate([1,2,3,1]))  # True
    print(containsDuplicate([1,2,3,4]))  # False
    print(containsDuplicate([1,1,1,3,3,4,3,2,4,2]))  # True
